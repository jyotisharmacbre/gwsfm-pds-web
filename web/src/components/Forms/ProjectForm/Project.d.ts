export interface Project {
  name: string;
  contractorId: number;
  companyId: number;
  headOfProject: string;
  projectManager: string;
  pmHasExperience: boolean;
  scope: string;
  cnNumber: number;
  status: number;
  engagementId: number;
  countryId: number;
  currencyId: number;
  probabilityOfWinning: String;
  approxValue: string;
  contractTypeId: string;
  cdmNotifiable: boolean;
  firstAssetWorkedOn: number;
  secondAssetWorkedOn: number;
  thirdAssetWorkedOn: number;
  comment: string;
  divisionId : string;
  businessUnitId: string;
}
