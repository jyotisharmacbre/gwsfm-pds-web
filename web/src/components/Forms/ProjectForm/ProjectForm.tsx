import React, { useEffect, useState } from 'react';
import { MainTitle } from '../../Title/Title';

import {
  Field,
  reduxForm,
  InjectedFormProps,
  formValueSelector
} from 'redux-form';
import PdsFormInput from '../../PdsFormHandlers/PdsFormInput';
import PdsFormSelect from '../../PdsFormHandlers/PdsFormSelect';
import PdsFormTextArea from '../../PdsFormHandlers/PdsFormTextArea';
import PdsFormButton from '../../PdsFormHandlers/PdsFormButton';
import { selectionButtons } from '../../../helpers/constants';
import {
  Validate,
  alphaNumeric,
  onlyNumber
} from '../../../helpers/fieldValidations';
import { connect } from 'react-redux';
import { IState } from '../../../store/state';
import { FormattedMessage, injectIntl } from 'react-intl';
import { LookupType } from '../../../store/Lookups/Types/LookupType';
import { getPropertyName,getDropdown, getFilterElementFromArray, normalizeToNumber } from '../../../helpers/utility-helper';
import PdsFormTypeAhead from '../../PdsFormHandlers/PdsFormTypeAhead';
import { IProjectDetail } from '../../../store/CustomerEnquiryForm/Types/IProjectDetail';
import { ICurrency } from '../../../store/Lookups/Types/ICurrency';
import Currency from '../../../store/Lookups/InitialState/Currency';
import IReactIntl from '../../../Translations/IReactIntl';
import TypeAhead from '../../TypeAhead/TypeAhead';
import { dynamicsContract } from '../../TypeAhead/TypeAheadConstantData/dynamicContractData';
import { dynamicsCompany } from '../../TypeAhead/TypeAheadConstantData/dynamicCompanyData';
import { IUserServiceData } from '../../../store/UserService/Types/IUserService';
import { any } from 'prop-types';

interface Props {
  projectstatus: any;
  onNext: (data: IProjectDetail) => void;
  onSave: (data: IProjectDetail) => void;
  currencies: Array<ICurrency> | null;
  onSearchContract: (value: any) => void;
  onSearchCompany: (value: any) => void;
  onSearchUserService: (value: any) => void;
  userServiceData: Array<IUserServiceData>;
}

const ProjectForm: React.FC<Props &
  IReactIntl &
  InjectedFormProps<IProjectDetail, Props>> = (props: any) => {
  const {
    handleSubmit,
    projectstatus,
    onSearchContract,
    onSearchCompany,
    onSearchUserService,
    userServiceData
  } = props;

  const otherDynamicsContract =
    props.dynamicsOtherContract.length > 0
      ? props.dynamicsOtherContract[0].label.split(' ')[0]
      : '';

  const otherDynamicsCompany =
    props.dynamicsOtherCompany.length > 0
      ? props.dynamicsOtherCompany[0].label
      : '';

      const getFormattedContractId = (customerId: string) => {
        return (customerId=== '0' ? '' : `(${customerId}), `)
      }
    
      const getFormattedCompanyId = (companyId: string) => {
        return (companyId=== '' ? '' : `(${companyId})`)
      }
    
   const getDynamicsContractDropdown =
      dynamicsContract &&
      dynamicsContract.map((ContractData: any) => {
        return { 
        label: `${ContractData.ContractName}${getFormattedContractId(ContractData.ContractId)}${ContractData.Name === '' ? '' : ContractData.Name}${getFormattedCompanyId(ContractData.CustomerId)}`,
        id: ContractData.ContractId }     
      }); 
    
  const getDynamicsCompanyDropdown =
  dynamicsCompany &&
  dynamicsCompany.map((CompanyData: any) => {
    return { label: CompanyData.Name, id: CompanyData.CompanyId };
  });
  const getUserServiceDropdown =
  userServiceData &&
  userServiceData.filter(user => user.firstname && user.lastName).map((UserServiceData: any) => {
    return { label: UserServiceData.firstname + " " + UserServiceData.lastName,
     id: UserServiceData.id,
      email: UserServiceData.email
    };
  });
  const CurrencyObj = new Currency();
  return (
    <div className="container-fluid">
      <div className=" row">
        <div className="col-lg-12 col-sm-12">
          <form className="customer-enquiry" onSubmit={handleSubmit}>
            <div className="row">
              <div className="col-lg-8">
                <MainTitle>
                  <FormattedMessage id="HEADING_CUSTOMER_ENQUIRY" />
                </MainTitle>
                <Field
                  name="name"
                  type="text"
                  component={PdsFormInput}
                  className="required"
                  validate={[
                    Validate.required('LABEL_PROJECT'),
                    Validate.maxLength(1000)
                  ]}
                  messageKey="MESSAGE_PROJECT_NAME"
                  labelKey="LABEL_PROJECT"
                  placeholderKey="PLACEHOLDER_PROJECT_NAME"
                />
                <TypeAhead name="companyId"
                options={getDynamicsCompanyDropdown}
                DynamicsType="companyId"
                onSearch={onSearchCompany}
                placeholderKey="PLACEHOLDER_COMPANY_NAME"
                className="required"
                labelName="LABEL_COMPANY"
                validationKey="LABEL_COMPANY"
                submitParam = "id"/>
                {otherDynamicsCompany === 'Other' && (
                  <Field
                    name="otherCompanyName"
                    type="text"
                    component={PdsFormInput}
                    className="required"
                    validate={[
                      Validate.required('LABEL_COMPANY'),
                      Validate.maxLength(1000)
                    ]}
                    labelKey="LABEL_OTHER_COMPANY"
                    placeholderKey="PLACEHOLDER_COMPANY_NAME"
                  />
                )}
                <TypeAhead name="contractorId"
                options={getDynamicsContractDropdown}
                DynamicsType="contractorId"
                onSearch={onSearchContract}
                placeholderKey="PLACEHOLDER_CONTRACT"
                className="required"
                labelName="LABEL_CONTRACT"
                validationKey="LABEL_CONTRACT"
                submitParam = "id"/>

                {otherDynamicsContract === 'Other' && (
                  <Field
                    name="otherContractName"
                    type="text"
                    component={PdsFormInput}
                    className="required"
                    validate={[
                      Validate.required('LABEL_CONTRACT'),
                      Validate.maxLength(1000)
                    ]}
                    labelKey="LABEL_OTHER_CONTRACT"
                    placeholderKey="PLACEHOLDER_CONTRACT"
                  />
                )}

<TypeAhead name="headOfProject"
                options={getUserServiceDropdown}
                DynamicsType="headOfProject"
                onSearch={onSearchUserService}
                placeholderKey="PLACEHOLDER_HEAD_OF_PROJECT_NAME"
                className="required"
                labelName="LABEL_HEAD_OF_PROJECT"
                validationKey="LABEL_HEAD_OF_PROJECT"
                submitParam = "email"/>

<TypeAhead name="projectOwner"
                options={getUserServiceDropdown}
                DynamicsType="projectOwner"
                onSearch={onSearchUserService}
                placeholderKey="PLACEHOLDER_PROJECT_OWNER_NAME"
                className="required"
                labelName="LABEL_PROJECT_OWNER"
                validationKey="LABEL_PROJECT_OWNER"
                submitParam = "email"/>


<TypeAhead name="projectManager"
                options={getUserServiceDropdown}
                DynamicsType="projectManager"
                onSearch={onSearchUserService}
                placeholderKey="PLACEHOLDER_PROJECT_MANAGER"
                className="required"
                labelName="LABEL_PROJECT_MANAGER"
                validationKey="LABEL_PROJECT_MANAGER"
                submitParam = "email"/>

                <Field
                  name="pmHasExperience"
                  component={PdsFormButton}
                  buttons={selectionButtons}
                  labelKey="LABEL_PROJECT_MANAGER_EXPERIENCEN"
                />

                <Field
                  name="scope"
                  rows="7"
                  component={PdsFormTextArea}
                  className="required"
                  validate={[
                    Validate.required('LABEL_PROJECT_SCOPE'),
                    Validate.maxLength(5000)
                  ]}             
                  labelKey="LABEL_PROJECT_SCOPE"
                />
                <Field
                  name="cnNumber"
                  type="text"
                  component={PdsFormInput}
                  validate={[
                    Validate.maxLength(25)
                  ]}  
                  labelKey="LABEL_CN_NUMBER"
                  placeholderKey="PLACEHOLDER_CN_NUMBER"
                />
                <div className={'form-group'}>
                  <label>
                    <FormattedMessage id="LABEL_PROJECT_STATUS" />
                  </label>
                  <div className="select-wrapper">
                    <Field
                      name="status"
                      component={PdsFormSelect}
                      placeHolder="Select status"
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_PROJECT_STATUS">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(
                        props.projectstatus,
                        LookupType.Project_Status
                      )}
                    </Field>
                  </div>
                </div>

                <div className={'form-group'}>
                  <label><FormattedMessage id='LABEL_TYPE_OF_ENGAGEMENT'></FormattedMessage></label>
                  <div className="select-wrapper">
                    <Field name="engagementId" component={PdsFormSelect} normalize={normalizeToNumber}>                    
                    <FormattedMessage id="PLACEHOLDER_TYPE_OF_ENGAGEMENT">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(
                        props.projectstatus,
                        LookupType.Engagement_Type
                      )}
                    </Field>
                  </div>
                </div>

                <div className={'form-group'}>
                  <label>
                    <FormattedMessage id="LABEL_COUNTRY" />*
                  </label>
                  <div className="select-wrapper">
                    <Field
                      name="countryId"
                      component={PdsFormSelect}
                      validate={Validate.required('LABEL_COUNTRY')}
                      placeholderKey="PLACEHOLDER_COUNTRY"
                      messageKey="MESSAGE_COUNTRY"
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_COUNTRY">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(props.projectstatus, LookupType.Country)}
                    </Field>
                  </div>
                </div>

                <div className={'form-group'}>
                  <label>
                    <FormattedMessage id="LABEL_CURRENCY" />*
                  </label>
                  <div className="select-wrapper">
                    <Field
                      name="currencyId"
                      component={PdsFormSelect}
                      validate={Validate.required('LABEL_CURRENCY')}
                      placeholderKey="PLACEHOLDER_CURRENCY"
                      messageKey="MESSAGE_CURRENCY"
                    >
                      <FormattedMessage id="PLACEHOLDER_CURRENCY">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {props.currencies &&
                        props.currencies.map((data: ICurrency, i: number) => {
                          return (
                            <option
                              key={data.currencyId}
                              value={data.currencyId}
                            >
                              {data.currencyName}
                            </option>
                          );
                        })}
                    </Field>
                  </div>
                </div>

                <Field
                  name="probabilityOfWinning"
                  type="number"
                  component={PdsFormInput}
                  labelKey="LABEL_PROBABILITY_OF_WINING"
                  placeholderKey="PLACEHOLDER_WIN_PROBABILITY"
                  className="width-100 required"
                  validate={[
                    Validate.required('LABEL_PROBABILITY_OF_WINING'),
                    Validate.maxLength(3),
                    onlyNumber
                  ]}
                  messageKey="MESSAGE_PROBABILITYOFWINING"
                />

                <Field
                  name="approxValue"
                  type="number"
                  component={PdsFormInput}
                  className="width-120 pl-20 required"
                  validate={[
                    Validate.required('LABEL_APPROXIMATE_VALUE'),
                    Validate.maxLength(1000),
                    onlyNumber
                  ]}
                  currency={getFilterElementFromArray(
                    props.currencies,
                    getPropertyName(
                    CurrencyObj,
                    prop => prop.currencyId
                  ),
                    props.currencyId,
                    getPropertyName(
                    CurrencyObj,
                    prop => prop.currencySymbol
                  )
                  )}
                  divPosition="relative"
                  labelKey="LABEL_APPROXIMATE_VALUE"
                  placeholderKey=""
                  messageKey="MESSAGE_APPROXIMATE_VALUE"
                />

                <div className={'form-group'}>
                  <label>
                    <FormattedMessage id="LABEL_CONTRACT_TYPE" />*
                  </label>
                  <div className="select-wrapper">
                    <Field
                      name="contractTypeId"
                      component={PdsFormSelect}
                      validate={Validate.required('LABEL_CONTRACT_TYPE')}
                      placeholderKey="PLACEHOLDER_CONTRACT_TYPE"
                      messageKey="MESSAGE_CONTRACT_TYPE"
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_CONTRACT_TYPE">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(
                        props.projectstatus,
                        LookupType.Contract_Type
                      )}
                    </Field>
                  </div>
                </div>

                <Field
                  name="cdmNotifiable"
                  component={PdsFormButton}
                  buttons={selectionButtons}
                  labelKey="LABEL_CDMNOTIFIABLE"
                />
                <div className={'form-group'}>
                  <label>
                    <FormattedMessage id="LABEL_ASSETS_WORKED_ON" />*
                  </label>
                  <div className="select-wrapper mb-2">
                    <Field
                      name="firstAssetWorkedOn"
                      component={PdsFormSelect}
                      DropdownCheck="selectRound"
                      placeholderKey="PLACEHOLDER_FIRST_ASSET"
                      messageKey="MESSAGE_FIRST_ASSET"
                      validate={[
                        Validate.required('LABEL_ASSETS_WORKED_ON'),
                        Validate.maxLength(1000),
                        onlyNumber
                      ]}
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_FIRST_ASSET">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      >{getDropdown(props.projectstatus, LookupType.Asset)}
                    </Field>
                  </div>

                  <div className="select-wrapper mb-2">
                    <Field
                      name="secondAssetWorkedOn"
                      component={PdsFormSelect}
                      DropdownCheck="selectRound"
                      placeholderKey="PLACEHOLDER_SECOND_ASSET"
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_SECOND_ASSET">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(props.projectstatus, LookupType.Asset)}
                    </Field>
                  </div>

                  <div className="select-wrapper mb-2">
                    <Field
                      name="thirdAssetWorkedOn"
                      component={PdsFormSelect}
                      DropdownCheck="selectRound"
                      placeholderKey="PLACEHOLDER_THIRD_ASSET"
                      normalize={normalizeToNumber}
                    >
                      <FormattedMessage id="PLACEHOLDER_THIRD_ASSET">
                        {message => <option value="">{message}</option>}
                      </FormattedMessage>
                      {getDropdown(props.projectstatus, LookupType.Asset)}
                    </Field>
                  </div>
                </div>

                <Field
                  labelKey="LABEL_COMMENTS"
                  name="comment"
                  rows="7"
                  component={PdsFormTextArea}
                  placeholderKey="PLACEHOLDER_ADDITIONAL_COMMENTS"
                />
              </div>
            </div>
            <div className="hr_line mb-0 mt-4"></div>
            <div className="mr-35 two-side-btn">
              <button
                className="active ml-auto"
                type="button"
                name="saveAndClose"
                onClick={handleSubmit(values => props.onSave(values))}
              >
                <FormattedMessage id="BUTTON_SAVE_AND_CLOSE" />
              </button>
              <button
                type="button"
                name="next"
                onClick={handleSubmit(values => props.onNext(values))}
              >
                <FormattedMessage id="BUTTON_NEXT" />
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};

const mapStateToProps = (state: IState) => ({
  initialValues: state.project.form,
  dynamicsOtherContract: state.dynamicData.dynamicsOtherContract,
  dynamicsOtherCompany: state.dynamicData.dynamicsOtherCompany,
  currencyId: selector(state, 'currencyId')
});

const form = reduxForm<IProjectDetail, Props>({
  form: 'ProjectForm',
  enableReinitialize: true
})(injectIntl(ProjectForm));

const selector = formValueSelector('ProjectForm');

export default connect(mapStateToProps)(form);
