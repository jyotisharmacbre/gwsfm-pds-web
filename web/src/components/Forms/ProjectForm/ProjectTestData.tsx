import Notify from '../../../enums/Notify';
import EventType from '../../../enums/EventType';
import { IDynamicsDivision, IDynamicBusinessUnits } from '../../../store/DynamicsData/Types/IDynamicData';

export const ProjectFormProps = {
    name: 'Test',
    contractorId: 1,
    companyId: 1,
    headOfProject: 'Test',
    projectOwner: 'Test',
    projectManager: 'Test',
    pmHasExperience: true,
    scope: 'Test',
    cnNumber: 2,
    status: 1,
    engagementId: 2,
    countryId: 3,
    currencyId: 1,
    probabilityOfWinning: 'Test',
    approxValue: 'Test',
    contractTypeId: 'Test',
    cdmNotifiable: true,
    firstAssetWorkedOn: 0,
    secondAssetWorkedOn: 0,
    thirdAssetWorkedOn: 0,
    comment: 'Test',
    divisionId: 'Test',
    businessUnitId: 'Test',
    weightedTCV: 10,
    soldMargin: 10
};

export const initialState = {
    form: {
        projectId: '',
        name: '',
        contractorId: 1,
        companyId: 1,
        headOfProject: '',
        projectOwner: '',
        projectManager: '',
        pmHasExperience: true,
        scope: '',
        cnNumber: 2,
        status: 1,
        engagementId: 2,
        countryId: 3,
        currencyId: 1,
        probabilityOfWinning: '',
        approxValue: '',
        contractTypeId: '',
        cdmNotifiable: true,
        firstAssetWorkedOn: 0,
        secondAssetWorkedOn: 0,
        thirdAssetWorkedOn: 0,
        comment: '',
        divisionId: '',
        businessUnitId: '',
        weightedTCV: 10,
        soldMargin: 10
    },
    enquiryOverview: {
        projectName: '',
        companyId: -1,
        headOfProject: '',
        projectManager: '',
        scope: '',
        cnNumber: -1
    },
    error: null,
    loading: false,
    notify: Notify.none,
    event: EventType.none,
    enquiryOverviewError: null
};

export const getprojectDetailData = {
    form: {
        projectId: '1',
        name: '',
        contractorId: 1,
        companyId: 1,
        headOfProject: '',
        projectOwner: '',
        projectManager: '',
        pmHasExperience: true,
        scope: '',
        cnNumber: 2,
        status: 1,
        engagementId: 2,
        countryId: 3,
        currencyId: 1,
        probabilityOfWinning: '',
        approxValue: '',
        contractTypeId: '',
        cdmNotifiable: true,
        firstAssetWorkedOn: 0,
        secondAssetWorkedOn: 0,
        thirdAssetWorkedOn: 0,
        comment: ''
    },
    enquiryOverview: {
        projectName: '',
        companyId: -1,
        headOfProject: '',
        projectManager: '',
        scope: '',
        cnNumber: -1
    },
    error: null,
    loading: false,
    notify: Notify.none,
    event: EventType.none,
    enquiryOverviewError: null
};

export const listOfDivisions: Array<IDynamicsDivision> = [{
    divisionId: 0,
    divisionCode: "0",
    description: "Other",
    dataSourceId: "0",
    isDeleted: "0"
},
{
    divisionId: 1,
    divisionCode: "A",
    description: "Other",
    dataSourceId: "0",
    isDeleted: "0"
}];
export const listOfBusinessUnits: Array<IDynamicBusinessUnits> = [{
    businessUnitId: "0",
    description: "Other",
    divisionCode: "0",
    divisionId: 0,
    isDeleted: "0"
}, {
    businessUnitId: "1",
    description: "AA",
    divisionCode: "0",
    divisionId: 1,
    isDeleted: "0"
}, {
    businessUnitId: "2",
    description: "BB",
    divisionCode: "0",
    divisionId: 1,
    isDeleted: "0"
    }];
