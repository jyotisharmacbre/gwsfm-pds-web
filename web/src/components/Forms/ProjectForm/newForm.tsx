import React from 'react';
import { connect } from 'react-redux';
import { Field, InjectedFormProps, reduxForm, change } from 'redux-form';
import ReduxFormInput from '../../components/ReduxFormInput';

const ProjectContainerForm = () => {
  return (
    <div className="container-fluid">
      <div className=" row">
        <div className="col-md-8">
          <form className="customer-enquiry">
            <h1>customer enquiry</h1>
            <div className="form-group">
              <label htmlFor="exampleInputEmail1">project*</label>
              <input
                type="email"
                className="form-control"
                placeholder="Enter project name"
              />
            </div>
          </form>
        </div>
      </div>
    </div>
  );
};
