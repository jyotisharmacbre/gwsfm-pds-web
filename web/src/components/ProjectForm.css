div.actions .leftalign{
   float: left;
}

div.actions .rightalign{
    float: right;
 }