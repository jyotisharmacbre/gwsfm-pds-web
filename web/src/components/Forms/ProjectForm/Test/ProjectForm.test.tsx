import React from 'react';
import { mount, ShallowWrapper } from 'enzyme';
import { Provider } from 'react-redux';
import { store } from '../../../../store';
import ProjectForm from '../ProjectForm';
import { IntlProvider } from 'react-intl';
import translations from '../../../../Translations/translation';
import * as connectedIntlProvider from './../../../../Translations/connectedIntlProvider';
import { IProjectDetailState } from '../../../../store/CustomerEnquiryForm/Types/IProjectDetailState';
import Notify from '../../../../enums/Notify';
import EventType from '../../../../enums/EventType';
import projectDetailReducer from '../../../../store/CustomerEnquiryForm/Reducer';
import { ActionType } from '../../../../store/CustomerEnquiryForm/Types/ActionType';
import nock from 'nock';
import { baseURL } from '../../../../client/client';
import { initialState, getprojectDetailData } from '../ProjectTestData';

nock(baseURL)
  .post('/api/Projects/customerEnquiry')
  .reply(200, "Project added successfully");

nock(baseURL)
  .put('/api/Projects/updatecustomerEnquiry')
  .reply(201, "Project updated successfully");

nock(baseURL)
  .put('/api/Projects/1/enquiryOverview')
  .reply(200, getprojectDetailData);


describe('ProjectForm Fields', () => {
  let wrapper: any;
  const props: any = {
    handleSubmit: jest.fn()
  };
  beforeEach(() => {
    const formatMessage = jest.mock('./../../../../Translations/connectedIntlProvider');

    jest.spyOn(connectedIntlProvider, 'formatMessage').mockImplementationOnce(() => {
      return 'intlmessage';
    });

    wrapper = mount(
      <Provider store={store}>
        <IntlProvider locale="en" messages={translations['en'].messages}>
          <ProjectForm {...props} />
        </IntlProvider>
      </Provider>
    );
  });
  it('Defines the component', () => {

    expect(wrapper).toBeDefined();
  });

  describe('Defines the Form', () => {
    let form: ShallowWrapper;
    beforeEach(() => {
      form = wrapper.find('[form="ProjectForm"]').first();
    });
    it('Renders form component', () => {

      expect(form).toHaveLength(1);
    });
  });

  describe('Defines form fields', () => {
    let field: ShallowWrapper;
    describe('Project name field', () => {
      beforeEach(() => {
        field = wrapper.find('input[name="name"]').first();
      });
      it('Should renders project name field', () => {
        expect(field.prop('type')).toBe('text');
      });
      it('Shows error when project name is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('ProbabilityOfWinning field', () => {
      beforeEach(() => {
        field = wrapper.find('input[name="probabilityOfWinning"]').first();
      });
      it('Should renders probabilityOfWinning field', () => {
        expect(field.prop('type')).toBe('number');
      });
      it('Shows error when probabilityOfWinning is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('ApproxValue field', () => {
      beforeEach(() => {
        field = wrapper.find('input[name="approxValue"]').first();
      });
      it('Should renders approxValue field', () => {
        expect(field.prop('type')).toBe('number');
      });
      it('Shows error when approxValue is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('ContractTypeId field', () => {
      beforeEach(() => {
        field = wrapper.find('select[name="contractTypeId"]').first();
      });
      it('Should renders contractTypeId field', () => {
        expect(field.render());
      });
      it('Shows error when contractTypeId is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('currencyId field', () => {
      beforeEach(() => {
        field = wrapper.find('select[name="currencyId"]').first();
        wrapper.setProps({
          currencies: [{currencyId: 1, currencySymbol: '$', currencyName: 'dollar'}],
          currencyId: 1
        });
      });
      it('Should renders currencyId field', () => {
        expect(field.render());
      });
      it('Shows error when currencyId is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('FirstAssetWorkedOn field', () => {
      beforeEach(() => {
        field = wrapper.find('select[name="firstAssetWorkedOn"]').first();
      });
      it('Should renders firstAssetWorkedOn field', () => {
        expect(field.render());
      });
      it('Shows error when firstAssetWorkedOn is set to blank', () => {
        field.simulate('blur');
        const errorBlock = wrapper.find('.text-danger');
        expect(errorBlock).toHaveLength(1);
      });
    });

    describe('Scope field', () => {
      beforeEach(() => {
        field = wrapper.find('textarea[name="scope"]').first();
      });
      it('Should renders scope field', () => {
        expect(field.render());
      });
    });

    describe('Comment field', () => {
      beforeEach(() => {
        field = wrapper.find('textarea[name="scope"]').first();
      });
      it('Should renders comment field', () => {
        expect(field.render());
      });
    });

    describe('Next button', () => {
      beforeEach(() => {
        field = wrapper.find('button[name="next"]').first();
      });
      it('Should renders next button', () => {
        expect(field.prop('type')).toBe('button');
      });
    });

    describe('SaveAndClose button', () => {
      beforeEach(() => {
        field = wrapper.find('button[name="saveAndClose"]').first();
      });
      it('Should renders saveAndClose button', () => {
        expect(field.prop('type')).toBe('button');
      });
    });

    describe('Typeahead field', () => {
      beforeEach(() => {
        field = wrapper.find('input[aria-autocomplete="both"]').first();
      });
      it('Should renders Typeahead field', () => {
        expect(field.render());
      });
    });

    describe('Project form reducer', () => {
      it('should handle get projectEnquiryOverview successfully', () => {
        const getProjectEnquiryOverviewAction: any = {
          type: ActionType.GET_ENQUIRY_OVERVIEW_SUCCESS,
          payload: { projectId: "1" }
        };
        expect(
          projectDetailReducer(initialState, getProjectEnquiryOverviewAction)
        ).toMatchSnapshot();
      });

      it('should handle edit project successfully', () => {
        const editProjectAction: any = {
          type: ActionType.PROJECT_EDIT_SUCCESS,
          payload: { projectId: "TestProjectId" }
        };
        expect(
          projectDetailReducer(initialState, editProjectAction)
        ).toMatchSnapshot();
      });

      it('should handle add project successfully', () => {
        const addProjectAction: any = {
          type: ActionType.PROJECT_ADD
        };
        expect(
          projectDetailReducer(initialState, addProjectAction)
        ).toMatchSnapshot();
      });

      it('should handle GET_PROJECT_DETAIL_ERROR with and return initialState', () => {
        const getProjectDetailError: any = {
          type: ActionType.GET_PROJECT_DETAIL_ERROR,
          error: null
        };
        expect(
          projectDetailReducer(initialState, getProjectDetailError)
        ).toMatchSnapshot();
      });
    });
  });
});
