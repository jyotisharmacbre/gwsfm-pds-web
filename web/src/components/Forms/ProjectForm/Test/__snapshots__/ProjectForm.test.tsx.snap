// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectForm Fields Defines form fields Project form reducer should handle GET_PROJECT_DETAIL_ERROR with and return initialState 1`] = `
Object {
  "enquiryOverview": Object {
    "cnNumber": -1,
    "companyId": -1,
    "headOfProject": "",
    "projectManager": "",
    "projectName": "",
    "scope": "",
  },
  "enquiryOverviewError": null,
  "error": null,
  "event": 3,
  "form": Object {
    "approxValue": "",
    "cdmNotifiable": true,
    "cnNumber": 2,
    "comment": "",
    "companyId": 1,
    "contractTypeId": "",
    "contractorId": 1,
    "countryId": 3,
    "currencyId": 1,
    "engagementId": 2,
    "firstAssetWorkedOn": 0,
    "headOfProject": "",
    "name": "",
    "pmHasExperience": true,
    "probabilityOfWinning": "",
    "projectId": "",
    "projectManager": "",
    "projectOwner": "",
    "scope": "",
    "secondAssetWorkedOn": 0,
    "status": 1,
    "thirdAssetWorkedOn": 0,
  },
  "loading": false,
  "notify": 0,
}
`;

exports[`ProjectForm Fields Defines form fields Project form reducer should handle add project successfully 1`] = `
Object {
  "enquiryOverview": Object {
    "cnNumber": -1,
    "companyId": -1,
    "headOfProject": "",
    "projectManager": "",
    "projectName": "",
    "scope": "",
  },
  "enquiryOverviewError": null,
  "error": null,
  "event": undefined,
  "form": Object {
    "approxValue": "",
    "cdmNotifiable": true,
    "cnNumber": 2,
    "comment": "",
    "companyId": 1,
    "contractTypeId": "",
    "contractorId": 1,
    "countryId": 3,
    "currencyId": 1,
    "engagementId": 2,
    "firstAssetWorkedOn": 0,
    "headOfProject": "",
    "name": "",
    "pmHasExperience": true,
    "probabilityOfWinning": "",
    "projectId": "",
    "projectManager": "",
    "projectOwner": "",
    "scope": "",
    "secondAssetWorkedOn": 0,
    "status": 1,
    "thirdAssetWorkedOn": 0,
  },
  "loading": false,
  "notify": 1,
}
`;

exports[`ProjectForm Fields Defines form fields Project form reducer should handle edit project successfully 1`] = `
Object {
  "enquiryOverview": Object {
    "cnNumber": -1,
    "companyId": -1,
    "headOfProject": "",
    "projectManager": "",
    "projectName": "",
    "scope": "",
  },
  "enquiryOverviewError": null,
  "error": null,
  "event": undefined,
  "form": Object {
    "approxValue": "",
    "cdmNotifiable": true,
    "cnNumber": 2,
    "comment": "",
    "companyId": 1,
    "contractTypeId": "",
    "contractorId": 1,
    "countryId": 3,
    "currencyId": 1,
    "engagementId": 2,
    "firstAssetWorkedOn": 0,
    "headOfProject": "",
    "name": "",
    "pmHasExperience": true,
    "probabilityOfWinning": "",
    "projectId": "TestProjectId",
    "projectManager": "",
    "projectOwner": "",
    "scope": "",
    "secondAssetWorkedOn": 0,
    "status": 1,
    "thirdAssetWorkedOn": 0,
  },
  "loading": false,
  "notify": 1,
}
`;

exports[`ProjectForm Fields Defines form fields Project form reducer should handle get projectEnquiryOverview successfully 1`] = `
Object {
  "enquiryOverview": Object {
    "projectId": "1",
  },
  "enquiryOverviewError": null,
  "error": null,
  "event": 3,
  "form": Object {
    "approxValue": "",
    "cdmNotifiable": true,
    "cnNumber": 2,
    "comment": "",
    "companyId": 1,
    "contractTypeId": "",
    "contractorId": 1,
    "countryId": 3,
    "currencyId": 1,
    "engagementId": 2,
    "firstAssetWorkedOn": 0,
    "headOfProject": "",
    "name": "",
    "pmHasExperience": true,
    "probabilityOfWinning": "",
    "projectId": "",
    "projectManager": "",
    "projectOwner": "",
    "scope": "",
    "secondAssetWorkedOn": 0,
    "status": 1,
    "thirdAssetWorkedOn": 0,
  },
  "loading": false,
  "notify": 0,
}
`;
